api/unstable/fingerprint.h

```diff
@@ -35,6 +35,102 @@ typedef enum {
     S2N_FINGERPRINT_JA3,
 } s2n_fingerprint_type;
 
+struct s2n_fingerprint;
+
+/**
+ * Create a reusable fingerprint structure.
+ *
+ * Fingerprinting is primarily used to identify malicious or abusive clients,
+ * so fingerprinting needs to be efficient and require minimal resources.
+ * The `s2n_client_hello_get_fingerprint_hash` and `s2n_client_hello_get_fingerprint_string`
+ * methods may require additional memory to calculate the fingerprint. Reusing
+ * the same `s2n_fingerprint` structure to calculate multiple fingerprints reduces
+ * the cost of each individual fingerprint.
+ *
+ * @param type The algorithm to use for the fingerprint.
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on failure.
+ */
+S2N_API struct s2n_fingerprint *s2n_fingerprint_new(s2n_fingerprint_type type);
+
+/**
+ * Frees the memory allocated by `s2n_fingerprint_new` for a fingerprint structure.
+ *
+ * @param fingerprint The fingerprint structure to be freed.
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on failure.
+ */
+S2N_API int s2n_fingerprint_free(struct s2n_fingerprint **fingerprint);
+
+/**
+ * Resets the fingerprint for safe reuse with a different ClientHello.
+ *
+ * @param fingerprint The fingerprint structure to be reset.
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on failure.
+ */
+S2N_API int s2n_fingerprint_reset(struct s2n_fingerprint *fingerprint);
+
+/**
+ * Sets the ClientHello to be fingerprinted.
+ *
+ * @param fingerprint The fingerprint to be modified
+ * @param ch The client hello to be fingerprinted. It will not be copied, so needs
+ * to live at least as long as this fingerprinting operation.
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on failure.
+ */
+S2N_API int s2n_fingerprint_set_client_hello(struct s2n_fingerprint *fingerprint, struct s2n_client_hello *ch);
+
+/**
+ * Get the size of the fingerprint hash.
+ *
+ * Fingerprint hashes should be a constant size, but that size will vary based
+ * on the fingerprinting method used.
+ *
+ * @param fingerprint The fingerprint to be used for the hash
+ * @param size Output variable to be set to the size of the hash
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on failure.
+ */
+S2N_API int s2n_fingerprint_get_hash_size(const struct s2n_fingerprint *fingerprint, uint32_t *size);
+
+/**
+ * Calculates a fingerprint hash.
```

Comment:
```suggestion
 * Calculates a hex-encoded fingerprint hash.
```

Reply:
Discussing in https://github.com/aws/s2n-tls/pull/4628#discussion_r1659064281